src/handle/why3_tactic.ml

```diff
@@ -42,30 +42,84 @@ let why3_env : Why3.Env.env =
+  ; symb_T     : sym (** Encoding of types.        *)
+  ; symb_or    : sym (** Disjunction(∨) symbol.    *)
+  ; symb_and   : sym (** Conjunction(∧) symbol.    *)
+  ; symb_imp   : sym (** Implication(⇒) symbol.    *)
+  ; symb_bot   : sym (** Bot(⊥) symbol.            *)
+  ; symb_top   : sym (** Top(⊤) symbol.            *)
+  ; symb_not   : sym (** Not(¬) symbol.            *)
+  ; symb_ex    : sym (** Exists(∃) symbol.         *)
+  ; symb_all   : sym (** Forall(∀) symbol.         *)
+  ; symb_codes : sym (** Type of encoded types.    *) }
 
 (** [get_config ss pos] build the configuration using [ss]. *)
 let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
   let builtin = Builtin.get ss pos in
-  { symb_P   = builtin "P"
-  ; symb_T   = builtin "T"
-  ; symb_or  = builtin "or"
-  ; symb_and = builtin "and"
-  ; symb_imp = builtin "imp"
-  ; symb_bot = builtin "bot"
-  ; symb_top = builtin "top"
-  ; symb_not = builtin "not" }
+  { symb_P     = builtin "P"
+  ; symb_T     = builtin "T"
+  ; symb_or    = builtin "or"
+  ; symb_and   = builtin "and"
+  ; symb_imp   = builtin "imp"
+  ; symb_bot   = builtin "bot"
+  ; symb_top   = builtin "top"
+  ; symb_not   = builtin "not"
+  ; symb_ex    = builtin "ex"
+  ; symb_all   = builtin "all"
+  ; symb_codes = builtin "Set" }
 
 (** A map from lambdapi terms to Why3 constants. *)
 type cnst_table = (term * Why3.Term.lsymbol) list
 
+(** Table of declared types, these types may be variables or constants. *)
+module TyTable : sig 
+  type t 
+
+  val empty : t 
+
+  val ty_of_term : t -> term -> (t * Why3.Ty.ty) option 
+  (** [ty_of_term tbl te] fetches the type associated to term [te] in
+      table [tbl] if it exists, or it creates such a type and return
+      the created type and the new table. *)
+
+  val fold_left : 
+    do_var:('a -> term -> Why3.Ty.tvsymbol -> 'a) ->
+    do_sym:('a -> term -> Why3.Ty.tysymbol -> 'a) -> 'a -> t -> 'a
+  (** [fold_left ~do_var ~do_sym init tbl] folds over the table [tbl]
+      with initial value [init] using [do_var] to operate on type
+      variables and [do_sym] to operate on type symbols. *)
+end = struct
+  (* The module is created to hide the type [sym_or_var] *)
+  type sym_or_var = TySym of Why3.Ty.tysymbol | TyVar of Why3.Ty.tvsymbol
+
+  type t = (term * sym_or_var) list
+
+  let empty = []
+
+  let ty_of_term tbl te =
+    match List.assoc_eq (Eval.eq_modulo []) te tbl with
+    | TySym s -> Some (tbl, Why3.Ty.ty_app s [])
+    | TyVar v -> Some (tbl, Why3.Ty.ty_var v)
+    | exception Not_found -> (
+        match get_args te with
+        | Symb s, [] ->
+            let sym = Why3.(Ty.create_tysymbol (Ident.id_fresh s.sym_name) [] Ty.NoDef) in
+            Some ((te,TySym sym)::tbl, Why3.Ty.ty_app sym [])
+        | Vari x, [] ->
+            let sym = Why3.(Ty.tv_of_string (Bindlib.name_of x)) in
+            Some ((te,TyVar sym)::tbl, Why3.Ty.ty_var sym)
+        | _ -> None)
```

Comment:
Why not returning a fresh type constant here, and make the function total (never returning None)?

Reply:
Right, I was not sure whether it was the right thing to do. I guess we want to have some conservativity property on the translation and that creating fresh constants does not break it.

Reply:
This is what we already do for propositions. It seems ok to me. If I remember well, there is a justification in Yacine's PhD.

Reply:
This is not ok in fact: I found a small bug in the correctness proof of the encoding of first-order logic (see my mail to deducteam). It can however be fixed by making sure that all type constants are inhabited.
I see two possible solutions to fix this problem:
- Translate only types (including constants) that we know are inhabited, making the translation partial. However, in LP, type inhabitation is undecidable. So, we could only rely on some necessarily incomplete heuristic. Moreover, I don't know how to handle type variables. More theoretical work on this topic is needed, that is, a correctness proof of the encoding of polymorphic first-order logic.
- A nice alternative would be to generate a fresh meta for each translated type, keeping the translation total, and provide a tactic implementing the inhabitation heuristic.

Reply:
Even without implementing talking about inhabitation heuristic, it would be already something to generate an inhabitation goal

Reply:
See https://www21.in.tum.de/~blanchet/tff1spec.pdf, page 7, for the definition of the semantics of polymorphic FOL.

Reply:
So what do you propose to do right now? To merge this PR as it is now but open an issue and fix it later in another PR?

Reply:
Ah why not, indeed we can do that.

Reply:
Or do you want to give it a try to generate metas?

Reply:
I can do that in a separate PR anyway